### PR TITLE
[DEV-635] change user

### PR DESCRIPTION
**Pull request by Krysh90**

### CODE REVIEW

File: app/components/ChangeUser.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+const ChangeUser = ({ user, onSubmit }: Props) => {
+  const { t } = useTranslation();
+  const [linkedAddresses, setLinkedAddresses] = useState<LinkedAddress[]>([]);
+
+  useEffect(() => {
+    const availableAddresses = user?.linkedAddresses ?? [];
+    setLinkedAddresses(availableAddresses);
+  }, []);
```
Comment (davidleomay) [inline app/components/ChangeUser.tsx:23]: Can this be moved to the default value?

File: app/components/ChangeUser.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    changeUser(linkedAddress)
+      .then((token) => SessionService.tokenLogin(token))
+      .catch(() => NotificationService.error(t("feedback.load_failed")))
+      .finally(onSubmit);
+  };
+
+  const textStyleFor = (linkedAddress: LinkedAddress, isAddress = false): TextStyle[] | undefined => {
+    const textStyles = [];
+    isAddress ? textStyles.push(styles.address) : textStyles.push(styles.blockchain);
```
Comment (davidleomay) [inline app/components/ChangeUser.tsx:41]: const textStyles = [isAddress ? styles.address : styles.blockchain];

File: app/components/ChangeUser.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+const ChangeUser = ({ user, onSubmit }: Props) => {
+  const { t } = useTranslation();
+  const [linkedAddresses, setLinkedAddresses] = useState<LinkedAddress[]>([]);
+
+  useEffect(() => {
+    const availableAddresses = user?.linkedAddresses ?? [];
+    setLinkedAddresses(availableAddresses);
+  }, []);
+
+  const changeTo = (linkedAddress?: LinkedAddress) => {
```
Comment (davidleomay) [inline app/components/ChangeUser.tsx:25]: How can the linkedAddress be undefined?

File: app/components/ChangeUser.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    if (!linkedAddress || linkedAddress.address === user?.address) return;
+
+    if (!linkedAddress.isSwitchable) {
+      NotificationService.warn(t("feedback.change_user_security"));
+      return;
+    }
+
+    changeUser(linkedAddress)
```
Comment (davidleomay) [inline app/components/ChangeUser.tsx:33]: Should we add some kind of a spinner/loader here?

File: app/components/ChangeUser.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import { LinkedAddress, UserDetail } from "../models/User";
+import { changeUser } from "../services/ApiService";
+import NotificationService from "../services/NotificationService";
+import SessionService from "../services/SessionService";
+import { Text, TouchableRipple } from "react-native-paper";
+import Colors from "../config/Colors";
+
+interface Props {
+  user?: UserDetail;
+  onSubmit: () => void;
```
Comment (davidleomay) [inline app/components/ChangeUser.tsx:13]: Why onSubmit? 😉

File: app/screens/LoginScreen.tsx [hunk 6]
```diff
@@ -206,17 +211,18 @@ const LoginScreen = () => {
...
+                      <Text>{signMessage}</Text>
                     </View>
                     <SpacerH />
                     <IconButton
                       icon={signCommandCopied ? "check" : "content-copy"}
                       color={signCommandCopied ? Colors.Success : Colors.Grey}
                       onPress={() => {
-                        ClipboardService.copy(signingCommand(address));
+                        if (!signMessage) return;
+                        ClipboardService.copy(signMessage);
```
Comment (ThaBrad) [inline app/screens/LoginScreen.tsx:225]: btw, an even more concise way would be: 
(signMessage) && ClipboardService.copy(signMessage) 😉

and the even safer condition would be 
(signMessage != null) && ClipboardService.copy(signMessage)

File: app/services/ApiService.ts [hunk 7]
```diff
@@ -64,6 +66,10 @@ export const signUp = (user: NewUser): Promise<string> => {
   return fetchFrom<AuthResponse>(`${AuthUrl}/signUp`, "POST", user).then((resp) => resp.accessToken);
 };
 
+export const getSignMessage = (address: string): Promise<ApiSignMessage> => {
+  return fetchFrom<ApiSignMessage>(`${AuthUrl}/signMessage?address=${address}`);
+};
+
```
Comment (ThaBrad) [inline app/services/ApiService.ts:72]: FYI
I added a (url) queryParam option in my implementation version, but not yet really needed now (just so you know where to look/copy if it's needed in the future 😉)
https://github.com/DFXswiss/wallet/pull/21/files#diff-0e865de905705adfe88bf0022e860df006ed1df41266d705f8597bcd2aa794c0

File: app/screens/LoginScreen.tsx (hunk 6 above)
Comment (ThaBrad) [reply app/screens/LoginScreen.tsx:225]: (early) returns, im allgemeinen, machen eigentlich nur bei Mehrzeiler wirklich Sinn

File: app/components/edit/BuyRouteEdit.tsx [hunk 8]
```diff
@@ -87,14 +92,34 @@ const BuyRouteEdit = ({
     iban: [Validations.Required, Validations.Iban(countries)],
   });
 
+  const getBuyTypes = (): BuyType[] => {
+    if (session?.blockchain === Blockchain.DEFICHAIN) return Object.values(BuyType);
+    return [BuyType.WALLET];
```
[the remark this answers is not in the document]
Comment (Krysh90) [reply app/components/edit/BuyRouteEdit.tsx]: or `session?.blockchain === undefined` => `Object.values(BuyType)`?

File: app/components/ChangeUser.tsx (hunk 3 above)
Comment (Krysh90) [reply app/components/ChangeUser.tsx:25]: true was maybe an older implementation. is now not necessary anymore

File: app/components/ChangeUser.tsx (hunk 5 above)
Comment (Krysh90) [reply app/components/ChangeUser.tsx:13]: no special reasoning, i can rename it to something else :D

File: app/screens/LoginScreen.tsx (hunk 6 above)
Comment (Krysh90) [reply app/screens/LoginScreen.tsx:228]: it is more than a one liner 🤔 As we want to show the checkmark only, if a sign message was copied

**Review by davidleomay — APPROVED** [review]

**Pull request merged**